### PR TITLE
Blog post on input checking

### LINKED ISSUE #149: Tools for input checking in functions
This follows a discussion on our institute's slack workspace with @pearsonca, @seabbs, @timtaylor. 

What are the best practices for input checking in functions? This is an important question since R doesn't have an explicit type system ([for](http://janvitek.org/pubs/oopsla20-r.pdf) [now](https://github.com/moodymudskipper/typed)?).

- the base solution `stop()`/`stopifnot()`. `stopifnot()` can also take named argument to produce nicer error messages. It's also possible to write your own function if you end up testing the same kind of things many times.
- [assertthat](https://github.com/hadley/assertthat) which seems to have been superseded by testthat (when used outside the testing framework)
- `vctrs::vec_assert()` which also works with `data.frame`s
- [checkmate](https://mllg.github.io/checkmate/index.html)
- [check](https://github.com/moodymudskipper/check)

check's README also mentions other alternatives that didn't come up in the discussion:
- https://bitbucket.org/richierocks/assertive
- https://docs.ropensci.org/assertr/
- https://github.com/smbache/ensurer

The popularity of each of these packages (as measured by the number of GitHub stars) makes a good case that such a blog post would be appreciated by the community :smile:

### PR DESCRIPTION
**Pull request by Bisaloo**
Fix #149.

All comments are welcome and I'm ready go through as many review rounds as necessary to reach the required quality for publication.

### CODE REVIEW

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,172 @@
+---
+slug: input-checking
+title: "Checking the inputs of your R functions" 
+authors: 
+- Sam Abbott
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:5]: Y'all will need author files cf https://github.com/r-hub/blog/tree/master/content/authors

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+
+say_hello("Bob")
+say_hello(404)
+```
+
+However, as you can see in this example, the error message is not in plain English but contains some code instructions.
+This can hinder understanding of the issue.
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:101]: Should https://style.tidyverse.org/error-messages.html be linked somewhere?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
+
+Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
+We will not get into the details of all of the available options here but below is a list of some of the them. If interested in understanding the various approaches to input taking the documentation
+for these package is a great place to start.
+
+-   testthat
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:158]: link all of them? Maybe `{testthat}` in the Rmd will do the trick.

Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:158]: what in testthat by the way?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+-   assertthat
+-   check
+-   assertr
+-   assertive
+-   ensurer
+-   `vctrs::vec_assert()`
+
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:165]: In the post I'd be curious to also hear about your motivation to write this and a short paragraph or so about what you ended up using in your package(s) and whether you got some user feedback.

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+## Introduction: the dangers of not checking function inputs
+
+R functions and R packages are a convenient way to share code with the rest of the world
+but it is generally not possible to know how, or with what precise aim in mind, others will
+use your code. For example, they might try to use it on objects that your function was not
+designed for. Let's imagine we have written a short function to compute the geometric mean:
+
+```{r}
+geometric_mean <- function(...) {
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:47]: this is a cool example!

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
+
+```{r}
+say_hello <- function(name) {
+  # Among other things, check_string() checks that we provide a 
+  # character object of length one
+  checkmate::assert_string(name)
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:139]: can the message be tweaked by the way? Isn't "assertion" a potentially unknown word for some users?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+```{r, error = TRUE}
+say_hello(404)
+```
+
+```{r, error = TRUE}
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:152]: I was also wondering whether, for some packages that are not testthat, the number of revdeps could be interesting?

**Review by maelle — CHANGES_REQUESTED** [review]
Thanks so much Hugo et al! Excellent input quality! :wink: I've made a few comments.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 3 above)
Comment (mpadge) [reply content/post/2021-10-07-input-checking/index.Rmd:158]: I agree - test that is out of place in this list, I'd suggest

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+-   assertive
+-   ensurer
+-   `vctrs::vec_assert()`
+
+## What about the future?
+
+In this post, we have discussed some methods to check function inputs, and to generate more informative error messages when doing so.
+However, this always comes with a performance cost, even though it's often relatively limited.
+Zero-cost assertions, as found in some other languages, would require some kind of typing system which R does not currently support.
+Interestingly several other languages have evolved to have typing systems as they have developed (TypeScript as an extension of JavaScript, type annotations in Python).
```
Comment (mpadge) [inline content/post/2021-10-07-input-checking/index.Rmd:171]: Thanks all for this great post! One very minor comment: That final sentence is crucial to understanding broader implications of what you've set out, and would work better without the refs to js and python in parentheses, and perhaps slightly tweaking to make more active. Maybe something like:
> ... typing systems as they have developed. Typescript developed as an extension of JavaScript, and type annotations are now possible in python. 

I like the final link! :smirk:

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 2 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:101]: It is linked just below, at line 109, as a footnote :wink:. Do you think it would be better here?

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 3 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:158]: My understanding is that assertthat was archived precisely because all its function was integrated in testthat, which is more actively developed.

It's a somewhat uncommon use case but you case use testthat outside of the testting context. E.g.,

``` r
test <- function(x) {
  
  testthat::expect_length(x, 1)
  
}

test(c(1,2))
#> Error: `x` has length 2, not length 1.
```

<sup>Created on 2022-01-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:158]: Your comments make me realise it may be nice to have a very small example for each package. So readers can get a feeling of how each package works. What do you think?

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 2 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:101]: Then maybe more prominent in the text? :sweat_smile: :woman_facepalming:

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 3 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:158]: If not too much work, yes that might be good, or a _link_ to an example in the wild e.g. from github.com/cran? E.g. are there many packages using testthat in this way?

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 2 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:101]: You mean I should promote the footnote to an actual sentence in the text?

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 3 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd]: Auto-linking with downlit (`{pkg}`) doesn't seem to work. I don't know why...

I manually added a link for each package.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 2 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:100]: yes, but you are free to disagree!

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 8 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd]: https://github.com/r-hub/blog/pull/150/commits/e5e9394c039f1b0b79c129b2799236a0250ea4cb

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 9]
```diff
@@ -168,5 +168,6 @@ If interested in understanding the various approaches to input taking the docume
 In this post, we have discussed some methods to check function inputs, and to generate more informative error messages when doing so.
 However, this always comes with a performance cost, even though it's often relatively limited.
 Zero-cost assertions, as found in some other languages, would require some kind of typing system which R does not currently support.
-Interestingly several other languages have evolved to have typing systems as they have developed (TypeScript as an extension of JavaScript, type annotations in Python).
+Interestingly several other languages have evolved to havetyping systems as they have developed.
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:171]:
```suggestion
Interestingly several other languages have evolved to have typing systems as they have developed.
```

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 6 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:139]: I don't think so. I think these packages have to be opinionated. If they leave too much flexibility to the user, we end up with `stopifnot()` or writing your own function.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 3 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd]: I'll remove testthat. I do think it has its place here but I agree it can be confusing because it's not the primary focus of this package.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 7 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:152]: Oh well, they are actually all on CRAN expected check. I might remove check (which is interesting but might have an overly complex syntax for a short demo) and add revdeps.

Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:152]: I keep going back and forth with this but they all have 0 revdeps, excepted asserthat, which has 11 revdeps. I think I will not include this.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 4 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:188]: In short: some of use were tired of filling our functions with `stopifnot()` / `if () { stop() }` or writing the same kind of checking function over and over and we started collectively looking for existing alternatives.

In which part of the post do you think this could fit?

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 2 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:100]: I do think it fits better as a footnote. This is not exactly part of the main topic of the post but more like an aside.

**Comment by pearsonca** [comment]
Pasting from internal slack conversation:

Overall seems fine. The particular issues that I don't see addressed (which reflect my somewhat curmudgeonly / luddite / high-expectations-for-users disposition):
 - absolutely first and foremost: document what the arguments are supposed to be. Telling the user what to provide is the first line of defense against them giving your code something stupid and first recourse to answer their questions about what went wrong when they still do.
 - I think there's more to say on the base R component. as argued in earlier slack thread: there's a balancing act here amongst competing s.e. principles: minimize dependencies, don't prematurely optimize, and DRY. Base R also offers other useful checking functions (`match.arg` e.g.). The appropriate error checking approach should be guided by what you're doing (both in the particular code and as a general habit)
 - think about the audience and what else you've done. again, if the arguments are documented with expectations (`#' @param name, a non-empty, non-NA character scalar`), then `stopifnot(missing(name), is.character(name), length(name) == 1, !is.na(name), name != "")` + the documentation is sufficient for general developers & sophisticated users, but maybe not novices and only-occasional-programmers
 - the fancy packages are mostly just convenience wrappers around the base R functionality to save you key strokes. that's fine and useful, but don't expect it to miraculously disappear complicated mutual constraints within and between complex input objects
  - in your closing paragraph, possibly worth noting that a big chunk of input validation here is because R doesn't have robust support of "type" like compiled languages (c++, e.g.). if your user-facing code is quite small, but there's lots of internal machinery that you're doing all sorts of input validation on to keep bugs from crawling in as you make changes, might be another note in favor of switching those internals to compiled code so that the compiler provides the guarantees

an aside probably not for the R audience, but in the more complex languages, the general problem of input validation is often called contracts or predicates and is being considered for sophisticated analysis even in older languages like c++ (https://www.reddit.com/r/cpp/comments/cmk7ek/what_happened_to_c20_contracts/ - though obviously not yet successfully) - though notably that concept also includes an element about the properties of results of a function, so is somewhat integrating unit-testing-like features more directly into code

**Comment by Bisaloo** [comment]
@pearsonca, I edited the text in accordance with your first two points. 

I appreciate your other points and think they contribute to an interesting discussion but I think they are slightly beyond the scope of this blog post. They are more opinionated when the current version of the post tries to present the various alternative without being too prescriptive. As you say, there is a balance to strike between somewhat competing principles and ultimately, the choice should depend on the context (are you in a situation where you particularly need to avoid dependencies? Who is your target audience? etc.). I feel confident that the readers will be able to decide where exactly to put the cursor depending on their needs and personal preferences.

I do think it is good if we can point out this discussion either by mentioning explicitly that readers could report to this PR for more details, or you could also post it again in the comments section of the blog after publication.

What do you think?

@maelle, what is your opinion on this?

**Comment by pearsonca** [comment]
Reference to PR makes sense to me.

**Comment by Bisaloo** [comment]
@TimTaylor, could you please review this and add an author file? https://github.com/r-hub/blog/tree/master/content/authors. I invited you as an external collaborator to the [fork in epiforecasts](https://github.com/epiforecasts/rhub_blog) but you can also post the content in a comment here.

File: content/authors/sam-abbott/index.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  link: https://samabbott.co.uk
+- service: Twitter
+  link: https://twitter.com/seabbs
+  icon: fa-twitter
+- service: Github
+  link: https://github.com/seabbs
+  icon: fa-github
+---
+
+Infectious disease researcher interested in open source tool development. More on my research interests [here](https://samabbott.co.uk/research).
```
Comment (maelle) [inline content/authors/sam-abbott/index.md:14]:
```suggestion
Infectious disease researcher interested in open source tool development. More on my [research interests](https://samabbott.co.uk/research).
```

https://webaccess.berkeley.edu/ask-pecan/click-here

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 4 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:236]: Maybe right before the introduction? Something like

> Are you, like we were, tired of filling your functions with blabla?

**Comment by maelle** [comment]
I'm reviewing the post but to answer

> @maelle, what is your opinion on this?

Maybe part of this could be worked into the conclusion?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+Or worse, it could give an incorrect output:
+
+```{r}
+geometric_mean(c(2, 8))
+```
+
+Because of this, you need to make sure you return clear errors whenever your functions receives input it was not designed for.
+In this blog post, we review a range of approaches to help you check your function inputs and discuss some potential future developments.
+
+## Pre-requisite: thouroughly document your argument types
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:83]:
```suggestion
## Pre-requisite: thoroughly document your argument types
```

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+#' @param name A character of length one with the name of the person to greet
+say_hello <- function(name) {
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+```
+
+Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.
+
+## Checking function inputs using base R
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:102]: Maybe also mention the rlang "equivalents"? https://rlang.r-lib.org/reference/index.html#check-arguments

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+[^3]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:181]:
```suggestion
checkmate provides a large number of functions that check that inputs respect a given set of properties, and that return clear error messages when that is not the case:
```

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+We will not get into the details of all of the available options here but below is a list of some of them, listed by decreasing number of reverse dependencies.
+If interested in understanding the various approaches to input taking the documentation for these package is a great place to start.
+
+-   [assertthat](https://github.com/hadley/assertthat)
+
+```{r, error = TRUE}
+assertthat::assert_that(is.character(1))
+```
+
+-   [assertr](https://docs.ropensci.org/assertr/)
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:212]: Interesting, I suppose it's an use case indeed, but I thought about it as an alternative to pointblank. :thinking:

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+```{r, error = TRUE}
+ensure_square <- ensurer::ensures_that(NCOL(.) == NROW(.))
+
+ensure_square(matrix(1:20, 4, 5))
+```
+
+-   `vctrs::vec_assert()`
+
+```{r, error = TRUE}
+vctrs::vec_assert(c(1, 2), "character")
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:238]: ça rend bizarre dans la preview (locale, `hugodown::hugo_start()`), y a-t-il une option à changer quand on l'utilise dans R Markdown ?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+
+vctrs::vec_assert(c(1, 2), size = 3)
+```
+
+-   [check](https://github.com/moodymudskipper/check) is slightly different because it doesn't provide utilities that work out of the box, but rather tools to assist you in writing your own checking functions
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:243]: Is there any use case to mention here?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+This way, you don't have to repeat yourself which is generally a good aim [^3].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^3]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:179]: Link https://blog.r-hub.io/2019/12/12/internal-functions/? :-)

**Review by maelle — COMMENTED** [review]
Very cool, thank you! Less comments this time :-)

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 15 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:238]: Maybe the `NO_COLOR` environment variable https://cli.r-lib.org/reference/faq.html

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+say_hello(404)
+```
+
+This is clearly a really great improvement to the functionality of base R.
+However, we can see from this example that we could create the error message programmatically based on the contents of the test.
+Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
+This way, you don't have to repeat yourself which is generally a good aim [^4].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^4]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:181]:
```suggestion
[^4]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](/2021/07/30/cache/)
```

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem [^5], you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of functions that check that inputs respect a given set of properties, and that return clear error messages when that is not the case:
+
+[^5]: See [this earlier blog post](https://blog.r-hub.io/2019/12/12/internal-functions/) for more information about why and who you would go with writing internal functions.
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:191]:
```suggestion
[^5]: See [this earlier blog post](/2019/12/12/internal-functions/) for more information about why and who you would go with writing internal functions.
```

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+  }
+)
+options(crayon.enabled = FALSE)
+
+```
+
+Are you, like we were, tired of filling your functions with argument checking code that sometimes ends up being longer that the core of the function itself?
+Are you trying to find what is the most efficient approach to check inputs easily and without forgetting any edge cases?
+Read about our exploration into the various ways to check your function inputs in R in this blog post.
+And please share your own tips and discoveries in the comment section below!
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:42]:
```suggestion
And please share your own tips and discoveries in the comment section!
```

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+```
+
+Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.
+
+## Checking function inputs using base R [^2]
+
+[^2]: Note that these base functions have equivalent in the tidyverse with a more consistent design and coloured output.
```
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:109]: In any case if you keep it as a footnote it'd be good if the footnote came _after_ explaining what the base R functions are.

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+```{r, error = TRUE}
+match.arg(arg = "R", choices = c("R", "python"))
+
+match.arg(arg = "javascript", choices = c("R", "python"))
+```
+
+But the real power of the `match.arg()` function comes from the fact that `choices` can be automatically obtained in the context of a function:
+
+```{r, error = TRUE}
+choose_language <- function(language = c("R", "python")) {
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:125]: One thing that's not obvious to me when using this pattern is what the default value is. Could you add a sentence about this?

File: content/post/2021-10-07-input-checking/index.Rmd [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+
+-   `vctrs::vec_assert()`
+
+```{r, error = TRUE}
+vctrs::vec_assert(c(1, 2), "character")
+
+vctrs::vec_assert(c(1, 2), size = 3)
+```
+
+-   [check](https://github.com/moodymudskipper/check) is slightly different because it doesn't provide utilities that work out of the box, but rather tools to assist you in writing your own checking functions
```
Comment (maelle) [inline content/post/2021-10-07-input-checking/index.Rmd:253]: Could you also add a link to a package using it?

**Review by maelle — COMMENTED** [review]
Thanks so much! I feel very nitpicky with my very last comments.

I really like the section about no approach being the best one.

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 22 above)
Comment (maelle) [reply content/post/2021-10-07-input-checking/index.Rmd:125]: (I know it _should_ be obvious to me :sweat_smile: :see_no_evil: )

**Comment by TimTaylor** [comment]
Hi @Bisaloo, apologies I've only had time to do a quick gander. If not too late in the day I think [{vetr}](https://cran.r-project.org/package=vetr) is definitely worth a mention in the alternatives. They also provide a good [comparison of different packages](http://htmlpreview.github.io/?https://github.com/brodieG/vetr/blob/master/extra/compare.html) which would be worth highlighting.

**Comment by pearsonca** [comment]
> Hi @Bisaloo, apologies I've only had time to do a quick gander. If not too late in the day I think [{vetr}](https://cran.r-project.org/package=vetr) is definitely worth a mention in the alternatives. They also provide a good [comparison of different packages](http://htmlpreview.github.io/?https://github.com/brodieG/vetr/blob/master/extra/compare.html) which would be worth highlighting.

@TimTaylor wow, thanks for highlighting `vetr`. I really like their concept; also seems to like it could dovetail well with unit testing practices - e.g. deriving template objects from the example objects used in testing (or vice versa).

File: content/post/2021-10-07-input-checking/index.Rmd (hunk 23 above)
Comment (Bisaloo) [reply content/post/2021-10-07-input-checking/index.Rmd:253]: I can't find any :/
